### PR TITLE
Update user avatar on navigation sidebar

### LINKED ISSUE #1757: The user avatar does not update, if avatar is uploaded from another device, unless user logout and login again.
1. Login into STG 'https://staging.fightpandemics.work' using valid credentials from Desktop
2. Navigate to user profile and upload the avatar.
3. Open FP - STG on the mobile, and user is already logged in on its mobile device with same account.

Issue: You will observe that the user avatar is not updated on the left navigation pop up, though it is already updated on the profile and post. User needs to logout and login again to sync the navigation sidebar and profile avatar. i.e. the session need to be reset.
Refer below GIF

ezgif com-video-to-gif 5

This issue exist for users who keeps the session active multiple devices. mainly on mobile as we have to display user avatar on profile page and left navigation sidebar.

### PR DESCRIPTION
**Pull request by Zaydme**
### What does this pull request aim to achieve? 💯
Makes the NavigationLayout have it's own UserContext so it can update the username/avatar and not use the global user object.
This update will be performed on route change.

_Please be concise and link any related issue(s):_
closes #1757 

- [x] I have checked that no one else is working on similar changes.
- [x] I have read the latest [**README**](README.md) and understand the development workflow.
- [x] The name of this branch is: **`feature/<branch_name>`**, or **`hotfix/<branch_name>`** if this is a hotfix, where `<branch_name>` briefly describes the issue(s) resolved and is prefixed with the issue number(s) (e.g. `1127-1130-update-git-branching-model`). The branch is created in a cloned version of this repo, **not a forked repo**.
- [x] This branch is rebased/merged with the **latest staging** (or the **latest production** if it is a hotfix).
- [x] There are no merge conflicts.
- [x] There are no console warnings and errors.
- [x] On the right hand side of this PR, I have linked the appropriate issue(s) under "Issues" and added a project progress under "Projects".
- [x] The title describes the issue(s) being addressed using format: **`<issue_numbers> - <brief_issue_description>`** (e.g. `#1127/#1130 - Update Git Branching Model`)

### CODE REVIEW

**Review by rabbi08 — APPROVED** [review]
Uploaded avatar from Desktop -> view on mobile, already opened application, just navigate to helpboard or so. the left navigation bar is updated along with profile pic with same avatar.

Uploaded avatar from Mobile-> view on Desktop, already opened application, just navigate to profile page or helpboard or other Org or so. the profile pic is updated as expected.

Thanks @Zaydme

**Comment by Zaydme** [comment]
This solution will work but it's not ideal, because it fetches the current user data on every route change, I don't know if this is fine.
But I can't think of any other way since we can't predict if the user was updated or not, so we just have to fetch it to find out.
Any thoughts @joshmorel ?

**Comment by galeza** [comment]
Hi,

I tested using: macOS and Android
Scenario 1: logged in on both devices
- upload avatar pic on desktop
- click Profile screen on Android - avatar was updated

Scenario 2: logged in on both devices
- upload new avatar on Android device
- click Help Board on ma macOS - avatar was updated for all my posts.

**Review by galeza — APPROVED** [review]
LGTM

File: client/src/templates/layouts/NavigationLayout.js [hunk 1]
```diff
@@ -288,6 +297,20 @@ const NavigationLayout = (props) => {
     { stateKey: "generalFeedback", label: t("feedback.otherFeedback") },
   ];
 
+  useEffect(() => {
```
Comment (joshmorel) [inline client/src/templates/layouts/NavigationLayout.js:300]: I can't recreate the issue in staging as described or I don't understand it...

But I agree with your point @Zaydme I don't think this is ideal. We shouldn't be fetching the user in this place just for this purpose...

In any case we will be changing navigation significantly with #1767 and then maybe again with some other redesign work so I don't think we should merge this until at least 1767 is done.

**Review by joshmorel — CHANGES_REQUESTED** [review]

File: client/src/templates/layouts/NavigationLayout.js (hunk 1 above)
Comment (joshmorel) [reply client/src/templates/layouts/NavigationLayout.js:300]: I'm caught up on the issue. I understand now.

File: client/src/templates/layouts/NavigationLayout.js [hunk 2]
```diff
@@ -263,10 +264,18 @@ const StyledDrawer = styled(Drawer)`
...
+    authLoading,
+    mobiletabs,
+    navSearch,
+    tabIndex,
+    isAuthenticated,
+  } = props;
   const history = useHistory();
   const [drawerOpened, setDrawerOpened] = useState(false);
   const [searchKeywords, setSearchKeywords] = useState(false);
+  const { userProfileState, userProfileDispatch } = useContext(UserContext);
```
Comment (joshmorel) [inline client/src/templates/layouts/NavigationLayout.js:277]:
```suggestion
```

But for the solution we should stick with `user` which comes from https://github.com/FightPandemics/FightPandemics/blob/1a471543dca12ccb00a7b63367ab3ef58356e2de/client/src/reducers/sessionReducer.js#L47. 

`UserContext` here is very confusing name wise, it should be called ProfileContext and shouldn't be used outside of org/user profile pages. Not sure when/why we diverted on state management but we definitely should be aiming for something more clear & consistent. A goal for post wide launch...

File: client/src/templates/layouts/NavigationLayout.js [hunk 3]
```diff
@@ -288,6 +297,20 @@ const NavigationLayout = (props) => {
     { stateKey: "generalFeedback", label: t("feedback.otherFeedback") },
   ];
 
+  useEffect(() => {
+    if (!isAuthenticated || !props.user) return;
+    (async function initialProfileFetch() {
```
Comment (joshmorel) [inline client/src/templates/layouts/NavigationLayout.js:302]:
```suggestion
    refetchUser();
```

Instead I'd dispatch this action (which will update the global user) similar to https://github.com/FightPandemics/FightPandemics/blob/1a471543dca12ccb00a7b63367ab3ef58356e2de/client/src/pages/ProfileCompleted.js#L23

**Review by joshmorel — CHANGES_REQUESTED** [review]
Now that I understand the problem (the videos caused confusion for me this time) I've added more specific suggestions.

However, another thing that could be solved here with some tweaks is to also "log them out" if they are no longer authenticated (e.g. logged out in other tab). Currently, the page needs to be refreshed for this to be clear. I think there might be an old issue related to that...

**Comment by joshmorel** [comment]
This is the related issue #782 (which I am still assigned :sweat_smile:).

I think we should aim to solve both these issues together as the mechanism for one should work for the other.

Either:

1. Similar to this resolution when they navigate the user is updated
2. Some kind of "global" heartbeat that queries the backend /api/users/current every 30-60s.

We'd just need to update `refreshUser` so if there's a 401 error thyen `dispatch({ type: AUTH_LOGOUT });`.

Or maybe modify `initAuth` function so it works for all uses?

**Comment by Zaydme** [comment]
> We'd just need to update refreshUser so if there's a 401 error thyen dispatch({ type: AUTH_LOGOUT });.

Or maybe not call the API at all if the auth cookie doesn't exist, and just dispatch logout.

Having an interval running every 30/60 seconds would probably make more requests than updating the user on route change, we just need to know how often the user changes the route, I am thinking that they will spend most of the time on the help board and casually checking some profiles, also when the tab is just open without any activity, maybe in the background, there will be no interval spamming the API. Maybe business intelligence can tell us how often the user changes the route on a single session.

**Comment by Zaydme** [comment]
@joshmorel 
Or even better, we can update the user on route change => store the time of the update, on the next route change, if it was less than 30 seconds ago don't refetch the user.
This way there will be no interval spamming the API, and nor quickly switching routes. So the user will be updated a maximum 1 time per 30 seconds and only when the user is actively using the website and not idle or in the background.

**Comment by joshmorel** [comment]
> Or maybe not call the API at all if the auth cookie doesn't exist, and just dispatch logout.

You are right. We could just this on a `setTimeout` loop and dispatch logout if it no longer exists: https://github.com/FightPandemics/FightPandemics/blob/431a96a3b8fcc8dce952ad132c5a962f1f40f85f/client/src/actions/authActions.js#L17

I've changed my mind about resolving them together I think we can solve them separately. The logout in other tab thing wouldn't really have an API usage implication then so maybe something that can be resolved without further discussion. 

The API usage implication but be worth it here but maybe not? I would see the need to login and logout to see updated avatar a bug, but not the need to reload the page (which seems to work fine). We can discuss further in Slack.

**Pull request closed without merging**